includes/Stories_Renderer/Renderer.php

```diff
@@ -0,0 +1,489 @@
+		 *
+		 * @param string $class Single story classes.
+		 */
+		return apply_filters( 'web_stories_renderer_single_story_classes', implode( ' ', $single_story_classes ) );
+	}
+
+	/**
+	 * Gets the container style attributes.
+	 *
+	 * @return string
+	 */
+	protected function get_container_styles() {
+
+		$container_style = '';
+
+		if ( true === $this->is_view_type( 'grid' ) ) {
+			$container_style = sprintf( 'grid-template-columns:repeat(%1$s, 1fr);', $this->attributes['number_of_columns'] );
+		}
+
+		/**
+		 * Filters the web stories renderer container style.
+		 *
+		 * @param string $class Container style.
+		 */
+		return apply_filters( 'web_stories_renderer_container_style', $container_style );
+	}
+
+	/**
+	 * Render story markup.
+	 *
+	 * @return void
+	 */
+	public function render_single_story_content() {
+
+		$story_data           = $this->get_story_item_data();
+		$single_story_classes = $this->get_single_story_classes();
+		$show_story_player    = ( true !== $this->attributes['show_story_poster'] && $this->is_view_type( 'grid' ) );
+
+		?>
+
+		<div class="<?php echo esc_attr( $single_story_classes ); ?>">
+			<?php
+
+			if ( true === $show_story_player ) {
+				$this->render_story_with_story_player( $story_data );
+			} else {
+				$this->render_story_with_poster( $story_data );
+			}
+			?>
+		</div>
+		<?php
+
+	}
+
+	/**
+	 * Renders a story with story's poster image.
+	 *
+	 * @param array $story_data Story item data. Contains information like url, height, width, etc of the story.
+	 *
+	 * @return void
+	 */
+	protected function render_story_with_poster( array $story_data ) {
+
+		$height                    = ( ! empty( $story_data['height'] ) ) ? absint( $story_data['height'] ) : 600;
+		$width                     = ( ! empty( $story_data['width'] ) ) ? absint( $story_data['width'] ) : 360;
+		$poster_style              = sprintf( 'background-image: url(%1$s);', esc_url_raw( $story_data['poster'] ) );
+		$list_view_image_alignment = '';
+
+		if ( true === $this->is_view_type( 'carousel' ) ) {
+			$poster_style = sprintf( '%1$s width: %2$spx; height: %3$spx', $poster_style, $width, $height );
+		}
+
+		if ( ! empty( $this->attributes['list_view_image_alignment'] ) ) {
+			$list_view_image_alignment = sprintf( 'image-align-%1$s', $this->attributes['list_view_image_alignment'] );
+		}
+
+		?>
+		<a class="<?php echo esc_attr( $list_view_image_alignment ); ?>"
+			href="<?php echo esc_url( $story_data['url'] ); ?>"
+		>
+			<div
+				class="web-stories-list__story-placeholder"
+				style="<?php echo esc_attr( $poster_style ); ?>"
+			></div>
+			<?php $this->get_content_overlay( $story_data ); ?>
+		</a>
+		<?php
+
+	}
+
+	/**
+	 * Renders a story with amp-story-player.
+	 *
+	 * @param array $story_data Story attributes. Contains information like url, height, width, etc of the story.
+	 *
+	 * @return void
+	 */
+	protected function render_story_with_story_player( array $story_data ) {
```

Comment:
Is this very similar to [this](https://github.com/google/web-stories-wp/blob/1431b73eb81051e98ddf60cedfcebce196cbd8ae/includes/Story_Renderer/Embed.php#L66-L136). Should make this more generatic so it can be reused.

Reply:
We can reuse that but in order to reuse we may need to refactor classes inside `includes/Story_Renderer`. Basically in ideal situation, we can get rid of those classes as the [Renderer interface](https://github.com/rtCamp/web-stories-wp/blob/rt/feature/php-api-stories/includes/Interfaces/Renderer.php) and [Base class](https://github.com/rtCamp/web-stories-wp/blob/rt/feature/php-api-stories/includes/Stories_Renderer/Renderer.php) is taking care of almost all these rendering logic.